app/services/sources/TwitterSource.java

```diff
@@ -1,17 +1,230 @@
+				.setOAuthAccessToken(env.get(TWITTER4J_ACCESSS_TOKEN))
+				.setOAuthAccessTokenSecret(env.get(TWITTER4J_ACCESS_TOKEN_SECRET));
+		TwitterFactory tf = new TwitterFactory(cb.build());
+		twitter = tf.getInstance();
+	}
+
+	/**
+	 * Gets tweets corresponding to the current trending topics on Twitter
+	 * @return
+	 */
+	public List<Post> getTrendingPosts(Trend trend, int numPosts) {
+		Query trendQuery = new Query(trend.getQuery());
+		trendQuery.setCount(numPosts);
+
+		try {
+			QueryResult result = twitter.search(trendQuery);
+			return parseQueryResult(result);
+		} catch (TwitterException e) {
+			e.printStackTrace();
+			// TODO
+		}
+
+		return Collections.emptyList();
+	}
+
+	public List<Post> getMaxTrendingPosts(Trend trend) {
+		return getTrendingPosts(trend, MAX_REQUEST_SIZE);
+	}
+
+	/**
+	 * Gets the {@link Trends} for a certain country, if it is available.
+	 * @param country
+	 * @param city
+	 * @return an {@link Optional} containing the trends, or an empty {@link Optional} if
+	 * {@link Trends} are not available for that country.
+	 */
+	public Optional<Trends> getTrends(String country, String city){
+
+		try{
+			Optional<Integer> id = getLocationId(country, city);
+
+			if (id.isPresent()) {
+				return Optional.of(twitter.getPlaceTrends(id.get()));
+			}
+		} catch (TwitterException e){
+			e.printStackTrace();
+			// TODO
+		}
+
+		return Optional.empty();
+	}
+
+	/**
+	 * Retrives the {@link Location} object for the specified country if it is available,
+	 * else an empty Optional.
+	 * @param country
+	 * @param city
+	 * @return {@link Optional} of the {@link Location}, or null if it is not available.
+	 */
+	private Optional<Integer> getLocationId(String country, String city) {
+		// check if location is cached
+		if (cachedCodes.containsKey(country)) {
+			Set<Location> cities = cachedCodes.get(country);
+
+			for (Location l : cities) {
+				if (l.getName().equalsIgnoreCase(city)) return Optional.of(l.getWoeid());
+			}
+		}
+
+		// else retrieve list of locations and store
+		try {
+			ResponseList<Location> availableLocations = twitter.getAvailableTrends();
+
+			Optional<Integer> toReturn = Optional.empty();
+			for (Location l : availableLocations) {
+				if (l.getCountryName().equalsIgnoreCase(country)
+						&& l.getName().equalsIgnoreCase(city)) {
+					toReturn = Optional.of(l.getWoeid());
+				}
+
+				if (cachedCodes.containsKey(l.getCountryCode())) {
+					cachedCodes.get(l.getCountryName()).add(l);
+				} else {
+					Set<Location> citySet = new HashSet<>();
+					citySet.add(l);
+					cachedCodes.put(l.getCountryName(), citySet);
+				}
+			}
+
+			return toReturn;
+		} catch (TwitterException e) {
+			e.printStackTrace();
+			// TODO
+		}
+
+		return Optional.empty(); // could not find the location
+	}
+
+	/**
+	 * Generates Post object for every status in given queried tweets
+	 * @param result
+	 * @return
+	 */
+	public List<Post> parseQueryResult(QueryResult result) {
```

Comment:
left it public because unimplemented test plans to use this. This is also a class that is meant to be used to do things so the high level methods being external seems like a good idea.